minor testing improvements

vtjnash opened the pull request.

vtjnash committed:
REPL test: fix destructive kill timer to do something

Previously, isopen(t) was guaranteed to be false, and the timer would just quietly expire

vtjnash committed:
avoid destroying error in try/finally of mktemp

it's very common for this `rm` to fail on Windows

vtjnash committed:
loading: fix variable name in debug message

vtjnash committed:
fix Pair method definition specialization

With `where {T}`, we will infer and cache this for `Type{T}`,
even though it's run for a `KindType`. Using `typeof` instead of an sparam
ensures that we don't need to fill the cache with those unused specializations.

vtjnash committed:
interrupt: fix race condition with errno

Caught by vchuravy!

It's likely that only the jl_safe_printf code to print
"force throwing a sigint" would alter errno anyways
(and thus, this fix may not be observable in practice),
but best to avoid altering `errno` in a signal handler regardless.

vtjnash committed:
Profile: eagerly drop some cache memory

We only need these `builder` fields during construction
(as a performance cache). As soon as that's done, we can drop them to
recover the memory (and it makes dump output on this tree nicer).

vtjnash committed:
CI: fix deprecation warning

ararslan commented on .circleci/config.yml:
```diff
@@ -50,7 +50,7 @@ jobs:
           cd .. &&
           mv project julia-src &&
-          /tmp/julia/bin/julia -e 'versioninfo()' &&
+          /tmp/julia/bin/julia -e 'Base.require(Main, :InteractiveUtils).versioninfo()' &&
```
Out of curiosity, why not simply `using InteractiveUtils; versioninfo()`?

staticfloat commented on stdlib/REPL/test/repl.jl:
```diff
@@ -16,8 +16,26 @@ include(joinpath(BASE_TEST_PATH, "testenv.jl"))
+        print(stderr, "WARNING: attempting hard kill of repl test after exceeding timeout\n")
+    end
+    return Timer(kill_test, delay)
+end
```
Instead of this `kill_timer`, I think it might make more sense for us to make a `readuntil_timed(io, needle; timeout=10)` or something.  That will allow us to create a closure within `readuntil` that would give us a useful backtrace, and would apply to all the `readuntil()` calls within this file, whereas this `kill_test()` doesn't apply to all of them (just most of them).

vtjnash replied:
This `kill_test` very aggressively applies to everything in this file (in a racy, panic-then-destroy-the-world-but-with-a-backtrace sort of way). As the comments above repeatedly say, the proper way to do this, in a place where you care about possible error recovery, is to close the `io` object.

Although maybe I should make the timeout much longer here. I've been doing some more testing, and it realized it could get pretty annoying to have your debugging session get killed if it lasts more than a couple minutes (as just happened to me). But then again, that's really just the price you have to pay for ever considering putting a time-based kill switch in your code.

StefanKarpinski replied:
Maybe don’t set the kill timer in an interactive session?

vtjnash replied:
AFAIK, we don't provide an equivalent of the Windows `IsDebuggerPresent` ccall for Unix.

vtjnash committed:
Serialization: fix a handful of implicit return mistakes

This would have caused us to perform a number of unncessary allocations and method specializations
